File: benchs/abrsort/abrsort.ml
type 'a tree = Empty | Node of 'a * 'a tree * 'a tree

let init n f =
  let rec aux n acc =
    if n = 0 then acc
    else
      aux (n-1) ((f n)::acc)
  in
  aux n []

let rec fold_left f accu l =
  match l with
    [] -> accu
  | a::l -> fold_left f (f accu a) l

let rec add a t =
  match t with
    | Empty -> Node (a,Empty, Empty)
    | Node (x,l,r) -> if a < x then Node(x,add a l,r) else Node(x,l,add a r)

let rec to_list t =
  match t with
    | Empty -> []
    | Node (x,l,r) -> List.rev_append (to_list l) (x :: to_list r)

let tree_sort l =
  let t = fold_left (fun acc x -> add x acc) Empty l in
  to_list t

let _ =
  let start = millis () in
  for i = 1 to 10000 do
    let l = init 100 (fun x -> Random.int x) in
    ignore (tree_sort l)
  done;
  let stop = millis () in
  Serial.write_string (string_of_int (stop-start))

File: benchs/abrsort/abrsort.py
import time
import random

class Node:

    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data

    def insert(self,item):
        if (item < self.data):
            if (self.left != None):
                self.left.insert(item)
            else:
                self.left = Node(item)
        else:
            if (self.right != None):
                self.right.insert(item)
            else:
                self.right = Node(item)

    def length(self):
        if self.data:
            if self.left is None:
                    ll = 0
            else:
                    ll = self.left.length()
            if self.right is None:
                    rr = 0
            else:
                    rr = self.right.length()
            return ll + rr + 1
        else:
            return 0

    def to_list(self):
        if self.data is None :
            return []
        else:
            if self.left is None:
                l = []
            else :
                l = self.left.to_list()
            if self.right is None:
                r = []
            else:
                r = self.right.to_list()
            l.append(self.data)
            l.extend(r)
            return l

    def display(self):
        if self.data is None :
            print (".")
        else :
            print(self.data)
            if self.left is not None :
                self.left.display()
            if self.right is not None :
                self.right.display()

def init(n,f):
    acc = []
    for i in range(0, n):
            acc.append(f(i))
    return acc

def tree_sort(l):
    i = l.pop(0)
    t = Node(i)
    for x in l:
        t.insert(x)
    t.display()
    return t.to_list()

def random_int(x):
    return random.randint(0,x)

def main():
    # start = time.ticks_ms()
    for i in range(0,10000):
        l = init(100,random_int)
    # stop=time.ticks_ms()
    # print(time.ticks_diff(stop,start))

if __name__ == "__main__":
    main()
